src/pgrepr/src/types.rs

```diff
@@ -84,6 +98,89 @@ pub enum Type {
     RegClass,
 }
 
+/// A length associated with [`Type::Char`] and [`Type::VarChar`].
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+pub struct CharLength(i32);
+
+impl CharLength {
+    fn from_typmod(typmod: i32) -> Option<CharLength> {
+        // https://github.com/postgres/postgres/blob/52377bb81/src/backend/utils/adt/varchar.c#L139
+        if typmod >= VARHDRSZ {
+            Some(CharLength(typmod - VARHDRSZ))
+        } else {
+            None
+        }
+    }
+
+    fn into_typmod(self) -> i32 {
+        // https://github.com/postgres/postgres/blob/52377bb81/src/backend/utils/adt/varchar.c#L60-L65
+        self.0 + VARHDRSZ
+    }
+
+    /// Consumes the newtype wrapper, returning the contents as an `i32`.
+    pub fn into_i32(self) -> i32 {
+        self.0
+    }
+}
+
+impl From<AdtCharLength> for CharLength {
+    fn from(length: AdtCharLength) -> CharLength {
+        // The `AdtCharLength` newtype wrapper ensures that the inner `u32` is
+        // small enough to fit into an `i32` with room for `VARHDRSZ`.
+        CharLength(i32::try_from(length.into_u32()).unwrap())
+    }
+}
+
+impl From<VarCharMaxLength> for CharLength {
+    fn from(length: VarCharMaxLength) -> CharLength {
+        // The `VarCharMaxLength` newtype wrapper ensures that the inner `u32`
+        // is small enough to fit into an `i32` with room for `VARHDRSZ`.
+        CharLength(i32::try_from(length.into_u32()).unwrap())
+    }
+}
+
+impl fmt::Display for CharLength {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        // https://github.com/postgres/postgres/blob/52377bb81/src/backend/utils/adt/varchar.c#L77
+        write!(f, "({})", self.0)
+    }
+}
+
+/// Constraints on [`Type::Numeric`].
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+pub struct NumericConstraints {
+    /// The maximum precision.
+    max_precision: i32,
+    /// The maximum scale.
+    max_scale: i32,
+}
+
+impl NumericConstraints {
+    fn from_typmod(typmod: i32) -> Option<NumericConstraints> {
```

Comment:
🤯 This seems like wants to share an implementation w/ https://github.com/MaterializeInc/materialize/blob/e49579ca8dd8e227e13bfd58df707d1ccf0e07b3/src/expr/src/scalar/func.rs#L5691, though maybe I'm being overeager

Once this PR lands, will see if I can find a place to tie these together.

Reply:
Oh 100% it should!! Good catch!